doc/man3/CRYPTO_THREAD_run_once.pod

```diff
@@ -98,6 +127,43 @@ read by CRYPTO_atomic_load() then CRYPTO_atomic_load() must be the only way that
 the variable is read. If atomic operations are not supported and I<lock> is
 NULL, then the function will fail.
 
+=item *
+
+CRYPTO_MUTEX_create() creates an architecture dependent mutex object.
+
+=item *
+
+CRYPTO_MUTEX_init() initializes a mutex object from CRYPTO_MUTEX_CREATE().
+CRYPTO_MUTEX_destroy() deinitializes the mutex and frees associated data
+
+=item *
+
+CRYPTO_MUTEX_lock() and CRYPTO_MUTEX_unlock() locks and unlocks the mutex,
+respectively.
+
+=item *
+
+CRYPTO_CONDVAR_create() creates a conditional variable. CRYPTO_CONDVAR_destroy()
+deinitializes frees associated data with the conditional variable
+
+=item *
+
+CRYPTO_CONDVAR_init() initializes a conditional variable.
+CRYPTO_CONDVAR_wait() expects an acquired mutex and awaits a broadcast.
+CRYPTO_CONDVAR_broadcast() executes a broadcast for a conditional variable.
+
+=item *
+
+CRYPTO_THREAD_enable() enables OpenSSL to spawn threads for a context.
+To impose an upper bound on the number of threads via I<max_threads>,
+or use -1 to impose no limit. CRYPTO_THREAD_cap() can be used to change
+the limit. To disable threads, use CRYPTO_THREAD_disable().
+CRYPTO_THREAD_enabled() indicates whether threading is enabled or not.
+
+CRYPTO_THREAD_spawn_worker() spawns a worker for use by OpenSSL. A callback
+I<cb> may be provided to to CRYPTO_THREAD_spawn_worker() to control the
+lifespan of a worker.
```

Comment:
Thank you very much for the feedback, comments inline:

> IIUC if we create the workers manually then OpenSSL will never spawn its own threads.

It may. Assume that OpenSSL may spawn threads `K` threads and it hasn't spawned any yet. Then it will spawn its own threads, if there are too few workers available (i.e., all busy or terminated). [Users provide an upper bound `K` on the number of threads, in which case OpenSSL may spawn (at any given time at most `K`, or infinity if `K = -1`) threads as needed.] IOW, the two can be combined.

> The whole point of having the externally spawned workers, IMO, is to enable the application to have control over thread creation and management.

Right now, the application may control the lifetime of a worker (thread) via a callback, but doesn't spawn it, nor does it manage taskqueue.

Do you think OpenSSL should implement that? The point behind externally spawned workers was to give users a bit more freedom beyond an upper limit (when they get spawned, when they get terminated), but not to force them to do task management whenever they wish to use threading (e.g., any Argon2 user). Achieving some of what you mention is not difficult (see below), but it would be nice to set some expectations of this API.

Since the "task/worker" and the "native" layers are separate, we can, instead of calling crypto_thread_native_start use a callback (if != NULL, otherwise default to crypto_thread_native_start) to start the thread instead (similarly we could create wrappers for join and terminate on the native layer).

> The external application might choose when and how to increase or decrease the thread pool size.

Increasing = may spawn new workers.

Decreasing can only be done via the worker callback/thread disable at the moment. Does it make sense to control it at any other time? Idle workers shouldn't matter much (?) and they get terminated on disable, busy workers cannot be terminated safely (only killed), and controlling a worker that finished a task is already implemented via a callback.

> Perhaps there might be some callbacks to tell the application when tasks have been added to the queue, or functions to find the task queue size.

We can call a callback when a task gets scheduled (i.e., not picked up by an active/new worker) in thread_schedule w/o too much difficulty.

> Finally there might be a function CRYPTO_THREAD_perform_thread_work() that an individual application managed thread could call to take a task off the queue and execute it.

It feels like this would force application developers to reinvent the wheel whenever they use threading. What's the motivation here?